gax/src/main/java/com/google/api/gax/batching/NonBlockingSemaphore.java

```diff
@@ -44,17 +45,27 @@ private static void checkNotNegative(long l) {
   NonBlockingSemaphore(long permits) {
     checkNotNegative(permits);
     this.currentPermits = new AtomicLong(permits);
+    this.limit = new AtomicLong(permits);
   }
 
+  @Override
   public void release(long permits) {
     checkNotNegative(permits);
-    currentPermits.addAndGet(permits);
+    while (true) {
+      long old = currentPermits.get();
+      if (permits + old > limit.get()) {
+        throw new IllegalStateException("Maximum permit count exceeded");
```

Comment:
similar to the blocking case, lets make this a saturating add and add a TODO come back to this